File: FrontEnd/lib/models/milestone.dart
class FundingMilestone {
  final int id;
  final int eventId;
  final String title;
  final String? description;
  final int unlockPercent;
  final String? benefitDescription;
  final int sortOrder;
  final int likeCount;
  final int dislikeCount;
  final bool isUnlocked;
  final DateTime createdAt;

  FundingMilestone({
    required this.id,
    required this.eventId,
    required this.title,
    this.description,
    required this.unlockPercent,
    this.benefitDescription,
    this.sortOrder = 0,
    this.likeCount = 0,
    this.dislikeCount = 0,
    this.isUnlocked = false,
    required this.createdAt,
  });

  factory FundingMilestone.fromJson(Map<String, dynamic> json) {
    return FundingMilestone(
      id: json['id'],
      eventId: json['event_id'],
      title: json['title'],
      description: json['description'],
      unlockPercent: json['unlock_percent'],
      benefitDescription: json['benefit_description'],
      sortOrder: json['sort_order'] ?? 0,
      likeCount: json['like_count'] ?? 0,
      dislikeCount: json['dislike_count'] ?? 0,
      isUnlocked: json['is_unlocked'] ?? false,
      createdAt: DateTime.parse(json['created_at']),
    );
  }
}

class MilestoneSnapshot {
  final int id;
  final int eventId;
  final int milestonePercent;
  final DateTime reachedAt;
  final int userCount;

  MilestoneSnapshot({
    required this.id,
    required this.eventId,
    required this.milestonePercent,
    required this.reachedAt,
    required this.userCount,
  });

  factory MilestoneSnapshot.fromJson(Map<String, dynamic> json) =>
      MilestoneSnapshot(
        id: json['id'] as int,
        eventId: json['event_id'] as int,
        milestonePercent: json['milestone_percent'] as int,
        reachedAt: DateTime.parse(json['reached_at'] as String),
        userCount: (json['user_count'] as int?) ?? 0,
      );
}

File: Backend/app/schemas/milestone.py
"""Funding milestone and early bird discount request/response schemas."""
from datetime import datetime

from pydantic import BaseModel, Field


class MilestoneCreate(BaseModel):
    title: str = Field(..., min_length=1, max_length=200)
    description: str | None = None
    unlock_percent: int = Field(..., ge=1, le=100)
    benefit_description: str | None = None
    sort_order: int = 0


class MilestoneUpdate(BaseModel):
    title: str | None = Field(None, min_length=1, max_length=200)
    description: str | None = None
    unlock_percent: int | None = Field(None, ge=1, le=100)
    benefit_description: str | None = None
    sort_order: int | None = None


class MilestoneResponse(BaseModel):
    id: int
    event_id: int
    title: str
    description: str | None = None
    unlock_percent: int
    benefit_description: str | None = None
    sort_order: int = 0
    like_count: int = 0
    dislike_count: int = 0
    is_unlocked: bool = False
    created_at: datetime

    model_config = {"from_attributes": True}


class MilestoneReactionResponse(BaseModel):
    action: str
    reaction: str
    like_count: int
    dislike_count: int


# ─── Milestone Snapshot Schemas ───


class MilestoneSnapshotResponse(BaseModel):
    id: int
    event_id: int
    milestone_percent: int
    reached_at: datetime
    user_count: int = 0

    model_config = {"from_attributes": True}


# ─── Early Bird Discount Schemas ───


class EarlyBirdDiscountCreate(BaseModel):
    applies_to: str = Field(..., pattern=r"^(funding|tickets)$")
    window_start: str | None = None  # ISO datetime (optional)
    window_end: str  # ISO datetime
    discount_type: str = Field(..., pattern=r"^(percent|fixed_cents)$")
    value: int = Field(..., ge=1)


class EarlyBirdDiscountUpdate(BaseModel):
    window_end: str | None = None
    discount_type: str | None = Field(None, pattern=r"^(percent|fixed_cents)$")
    value: int | None = Field(None, ge=1)


class EarlyBirdDiscountResponse(BaseModel):
    id: int
    event_id: int
    applies_to: str
    window_start: datetime | None = None
    window_end: datetime
    discount_type: str
    value: int
    is_active: bool = False
    created_at: datetime

    model_config = {"from_attributes": True}
